lib/ast.js

```diff
@@ -91,7 +91,15 @@ var AST_Token = DEFNODE("Token", "type value line col pos endline endcol endpos
 }, null);
 
 var AST_Node = DEFNODE("Node", "start end", {
-    clone: function() {
+    clone: function(deep) {
+        if (deep) {
+            var self = this.clone();
+            return self.transform(new TreeTransformer(function(node) {
+                if (node !== self) {
+                    return node.clone(true);
```

Comment:
Those are covered:

https://github.com/mishoo/UglifyJS2/blob/f4c2ea37bf9231b6f76804e74ee157be916280de/lib/transform.js#L168

https://github.com/mishoo/UglifyJS2/blob/f4c2ea37bf9231b6f76804e74ee157be916280de/lib/transform.js#L79

https://github.com/mishoo/UglifyJS2/blob/f4c2ea37bf9231b6f76804e74ee157be916280de/lib/utils.js#L121

https://github.com/mishoo/UglifyJS2/blob/f4c2ea37bf9231b6f76804e74ee157be916280de/lib/utils.js#L155

So those sites should have a fresh copy of array with cloned nodes.